File: services/frontend/src/types/DeathSurvivor.ts
/**
 * TypeScript types for the Death & Survivor Benefits domain.
 * Defines interfaces for death records, survivor claims, death benefit
 * elections, overpayment tracking, and the complete death processing summary.
 *
 * Consumed by: death-survivor-demo-data.ts, death stage components
 * Depends on: Member.ts (AuditEntry type)
 */

/** Death record status tracking — follows the death processing workflow */
export type DeathRecordStatus = 'NOTIFIED' | 'VERIFIED' | 'PROCESSED' | 'CLOSED'

/** Survivor claim types */
export type SurvivorClaimType = 'JS_SURVIVOR' | 'CONTRIB_REFUND' | 'SURVIVOR_ANNUITY'

/** Survivor claim status */
export type SurvivorClaimStatus = 'PENDING' | 'IN_REVIEW' | 'APPROVED' | 'ACTIVE' | 'DENIED' | 'CLOSED'

/** Overpayment recovery status */
export type OverpaymentRecoveryStatus = 'IDENTIFIED' | 'REQUESTED' | 'PARTIAL' | 'RECOVERED' | 'WAIVED' | 'UNCOLLECTABLE'

/** Death record for a member — created upon notification of death */
export interface DeathRecord {
  death_record_id: number
  member_id: string
  death_date: string
  notification_date: string
  notification_source: string
  notification_contact?: string
  certificate_received_date?: string
  certificate_verified_date?: string
  certificate_verified_by?: string
  previous_member_status: string
  current_member_status: string
  status: DeathRecordStatus
  suspend_date?: string
  final_payment_date?: string
  overpayment_detected: boolean
  overpayment_amount: number
}

/** Survivor benefit claim — filed by designated survivor or beneficiary */
export interface SurvivorClaim {
  claim_id: number
  death_record_id: number
  member_id: string
  survivor_first_name: string
  survivor_last_name: string
  survivor_date_of_birth?: string
  survivor_relationship: string
  claim_type: SurvivorClaimType
  js_percentage?: number
  monthly_amount: number
  lump_sum_amount?: number
  effective_date: string
  first_payment_date?: string
  status: SurvivorClaimStatus
  approved_date?: string
  approved_by?: string
}

/** Death benefit installment election — made at retirement */
export interface DeathBenefitElection {
  election_id: number
  member_id: string
  lump_sum_amount: number
  num_installments: number
  installment_amount: number
  effective_date: string
  installments_paid: number
  installments_remaining: number
  remaining_amount: number
  beneficiary_first_name?: string
  beneficiary_last_name?: string
  beneficiary_relationship?: string
  status: string
}

/** Overpayment record — a single post-death overpayment */
export interface OverpaymentRecord {
  overpayment_id: number
  member_id: string
  payment_date: string
  payment_amount: number
  recovery_status: OverpaymentRecoveryStatus
  recovery_amount: number
  recovery_source?: string
}

/** Death benefit status summary — aggregates all death processing info */
export interface DeathBenefitStatus {
  member_id: string
  has_death_record: boolean
  death_record?: DeathRecord
  survivor_claims: SurvivorClaim[]
  death_benefit_election?: DeathBenefitElection
  overpayment_records: OverpaymentRecord[]
  processing_complete: boolean
}

/** Overpayment detection summary from the intelligence engine */
export interface OverpaymentInfo {
  overpayment_count: number
  overpayment_total: number
  valid_payments: number
  payment_details: Array<{
    deposit_date: string
    amount: number
    valid: boolean
    reason: string
  }>
}

/** Survivor benefit calculation result from intelligence engine */
export interface SurvivorBenefitResult {
  survivor_monthly_benefit: number
  survivor_name: string
  js_percentage: number
  effective_date: string
  duration: string
  formula: string
}

/** Death benefit installment calculation result */
export interface InstallmentCalcResult {
  installment_amount: number
  installments_paid: number
  installments_remaining: number
  remaining_total: number
  formula: string
}

/** Active member death determination result */
export interface ActiveMemberDeathResult {
  benefit_type: 'contribution_refund' | 'survivor_annuity'
  vested: boolean
  refund_amount?: number
  survivor_annuity_available: boolean
  formula: string
}

/** Pop-up provision result */
export interface PopUpResult {
  new_benefit: number
  increase_amount: number
  effective_date: string
  new_beneficiary_allowed: boolean
  retroactive: boolean
  formula: string
}

/** Complete death processing summary — the full package from the intelligence engine */
export interface DeathProcessingSummary {
  member_id: string
  notification: {
    benefit_suspended: boolean
    certificate_required: boolean
    status_transition: string
    note: string
  }
  overpayment: OverpaymentInfo
  survivor_benefit?: SurvivorBenefitResult
  death_benefit_installments?: InstallmentCalcResult
  active_member_death?: ActiveMemberDeathResult
  record_transition: {
    status_sequence: string[]
    survivor_record_created: boolean
    benefit_terminated: boolean
  }
  calculation_trace: Array<{
    step: number
    rule_id: string
    rule_name: string
    description: string
    inputs: string
    result: string
    source_reference: string
  }>
}

File: services/connector/internal/models/death_survivor.go
// Package models — death and survivor benefit domain types.
// Maps DEATH_RECORD, SURVIVOR_CLAIM, DEATH_BENEFIT_ELECTION, and OVERPAYMENT_RECORD
// from 004_death_survivor_schema.sql to clean Go structs.
//
// Consumed by: death_queries.go (DB access), death_handlers.go (API layer)
// Depends on: 004_death_survivor_schema.sql table definitions
package models

import "time"

// DeathRecord represents a member death processing record from DEATH_RECORD.
// Created upon credible notification of a member's death. Tracks the entire
// death processing workflow from notification through final status transition.
type DeathRecord struct {
	DeathRecordID   int        `json:"death_record_id"`
	MemberID        string     `json:"member_id"`
	DeathDate       *time.Time `json:"death_date,omitempty"`
	NotifyDate      time.Time  `json:"notification_date"`
	NotifySource    string     `json:"notification_source,omitempty"`
	NotifyContact   string     `json:"notification_contact,omitempty"`
	NotifyPhone     string     `json:"notification_phone,omitempty"`
	CertReceivedDate *time.Time `json:"certificate_received_date,omitempty"`
	CertVerifyDate  *time.Time `json:"certificate_verified_date,omitempty"`
	CertVerifyBy    string     `json:"certificate_verified_by,omitempty"`
	CertDocRef      string     `json:"certificate_document_ref,omitempty"`
	PreviousStatus  string     `json:"previous_member_status,omitempty"`
	CurrentStatus   string     `json:"current_member_status,omitempty"`
	StatusCode      string     `json:"status"`
	SuspendDate     *time.Time `json:"suspend_date,omitempty"`
	FinalPayDate    *time.Time `json:"final_payment_date,omitempty"`
	OverpayFlag     string     `json:"overpayment_detected"`
	OverpayAmount   float64    `json:"overpayment_amount,omitempty"`
	Notes           string     `json:"notes,omitempty"`
}

// SurvivorClaim represents a survivor benefit claim from SURVIVOR_CLAIM.
// Created when a designated survivor or beneficiary files for benefits
// following a member's death.
type SurvivorClaim struct {
	ClaimID         int        `json:"claim_id"`
	DeathRecordID   int        `json:"death_record_id"`
	MemberID        string     `json:"member_id"`
	SurvivorFirst   string     `json:"survivor_first_name,omitempty"`
	SurvivorLast    string     `json:"survivor_last_name,omitempty"`
	SurvivorDOB     *time.Time `json:"survivor_date_of_birth,omitempty"`
	SurvivorRelation string    `json:"survivor_relationship,omitempty"`
	ClaimType       string     `json:"claim_type"`
	JSPercentage    *float64   `json:"js_percentage,omitempty"`
	MonthlyAmount   float64    `json:"monthly_amount,omitempty"`
	LumpSumAmount   float64    `json:"lump_sum_amount,omitempty"`
	EffectiveDate   *time.Time `json:"effective_date,omitempty"`
	FirstPayDate    *time.Time `json:"first_payment_date,omitempty"`
	StatusCode      string     `json:"status"`
	ApprovedDate    *time.Time `json:"approved_date,omitempty"`
	ApprovedBy      string     `json:"approved_by,omitempty"`
	DenyReason      string     `json:"deny_reason,omitempty"`
	Notes           string     `json:"notes,omitempty"`
}

// DeathBenefitElection represents the death benefit installment election
// from DEATH_BENEFIT_ELECTION. Elected at retirement (50 or 100 installments).
// After member death, remaining installments transfer to the designated beneficiary.
type DeathBenefitElection struct {
	ElectionID        int        `json:"election_id"`
	MemberID          string     `json:"member_id"`
	LumpSumAmount     float64    `json:"lump_sum_amount"`
	NumInstallments   int        `json:"num_installments"`
	InstallmentAmount float64    `json:"installment_amount"`
	EffectiveDate     time.Time  `json:"effective_date"`
	InstallmentsPaid  int        `json:"installments_paid"`
	RemainingAmount   float64    `json:"remaining_amount"`
	BeneficiaryFirst  string     `json:"beneficiary_first_name,omitempty"`
	BeneficiaryLast   string     `json:"beneficiary_last_name,omitempty"`
	BeneficiaryRelation string   `json:"beneficiary_relationship,omitempty"`
	TransferDate      *time.Time `json:"transfer_date,omitempty"`
	StatusCode        string     `json:"status"`
}

// OverpaymentRecord represents a post-death overpayment from OVERPAYMENT_RECORD.
// Each record tracks a single overpaid payment and its recovery status.
type OverpaymentRecord struct {
	OverpaymentID   int        `json:"overpayment_id"`
	DeathRecordID   int        `json:"death_record_id"`
	MemberID        string     `json:"member_id"`
	PaymentDate     time.Time  `json:"payment_date"`
	PaymentAmount   float64    `json:"payment_amount"`
	RecoveryStatus  string     `json:"recovery_status"`
	RecoveryAmount  float64    `json:"recovery_amount"`
	RecoveryDate    *time.Time `json:"recovery_date,omitempty"`
	RecoverySource  string     `json:"recovery_source,omitempty"`
	Notes           string     `json:"notes,omitempty"`
}

// DeathNotificationRequest is the input for POST /api/v1/members/{id}/death/notify.
type DeathNotificationRequest struct {
	MemberID       string `json:"member_id"`
	DeathDate      string `json:"death_date,omitempty"`
	NotifySource   string `json:"notification_source"`
	NotifyContact  string `json:"notification_contact,omitempty"`
	NotifyPhone    string `json:"notification_phone,omitempty"`
	Notes          string `json:"notes,omitempty"`
}

// SurvivorClaimRequest is the input for POST /api/v1/members/{id}/survivor-claims.
type SurvivorClaimRequest struct {
	MemberID         string  `json:"member_id"`
	SurvivorFirst    string  `json:"survivor_first_name"`
	SurvivorLast     string  `json:"survivor_last_name"`
	SurvivorDOB      string  `json:"survivor_date_of_birth,omitempty"`
	SurvivorRelation string  `json:"survivor_relationship"`
	ClaimType        string  `json:"claim_type"`
	JSPercentage     float64 `json:"js_percentage,omitempty"`
}
